File: 111. Minimum Depth of Binary Tree/minDepth.py
from typing import Optional
# Definition for a binary tree node.
class TreeNode:
  def __init__(self, val=0, left=None, right=None):
    self.val = val
    self.left = left
    self.right = right

class Solution:
  def minDepth(self, root: Optional[TreeNode]) -> int:
    def dfs(node):
      if not node:
        return 0
      
      if not node.left:
        return dfs(node.right) + 1
      
      if not node.right:
        return dfs(node.left) + 1

      return min(dfs(node.left), dfs(node.right)) + 1
    
    return dfs(root)


root = TreeNode(3)
root.left = TreeNode(9)
root.right = TreeNode(20)

root.right.left = TreeNode(15)
root.right.right = TreeNode(7)

solution = Solution()
print('minDepth: ', solution.minDepth(root))

File: 111. Minimum Depth of Binary Tree/minDepth.js
// Definition for a binary tree node.
class TreeNode {
  constructor(val = 0, left = null, right = null) {
    this.val = val;
    this.left = left;
    this.right = right;
  }
}
/**
 * @param {TreeNode} root
 * @return {number}
 */
const minDepth = (root) => {
  const dfs = (node) => {
    if (!node) {
      return 0;
    }

    if (!node.left) {
      return dfs(node.right) + 1;
    }

    if (!node.right) {
      return dfs(node.left) + 1;
    }

    return Math.min(dfs(node.left), dfs(node.right)) + 1; 
  }

  return dfs(root);
};


const root = new TreeNode(3);
root.left = new TreeNode(9);
root.right = new TreeNode(20);

root.right.left = new TreeNode(15);
root.right.right = new TreeNode(7);

console.log('minDepth: ', minDepth(root))
